silkworm/api/silkworm_api.h

```diff
@@ -0,0 +1,88 @@
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#ifndef SILKWORM_API_H_
+#define SILKWORM_API_H_
+
+// C API exported by Silkworm to be used in Erigon.
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+#pragma GCC diagnostic ignored "-Wsign-conversion"
+#include <mdbx.h>
+#pragma GCC diagnostic pop
+#include <stdbool.h>
+#include <stdint.h>
+
+#if defined _MSC_VER
+#define SILKWORM_EXPORT __declspec(dllexport)
+#else
+#define SILKWORM_EXPORT __attribute__((visibility("default")))
+#endif
+
+#if __cplusplus
+#define SILKWORM_NOEXCEPT noexcept
+#else
+#define SILKWORM_NOEXCEPT
+#endif
+
+#if __cplusplus
+extern "C" {
+#endif
+
+enum SilkwormStatusCode {
+    kSilkwormSuccess = 0,
+    kSilkwormInvalidBlockRange = 1,
+    kSilkwormBlockNotFound = 2,
+    kSilkwormUnknownChainId = 3,
+    kSilkwormMdbxError = 4,
+    kSilkwormInvalidBlock = 5,
+    kSilkwormDecodingError = 6,
+    kSilkwormUnknownError = -1
+};
+
+/** @brief Executes a batch of Ethereum blocks and writes resulting changes into the database.
+ *
+ * @param[in] txn Valid read-write MDBX transaction. Must not be NULL.
+ * This function does not commit nor abort the transaction.
+ * @param[in] chain_id EIP-155 chain ID. kSilkwormUnknownChainId is returned in case of an unknown or unsupported chain.
+ * @param[in] start_block The block height to start the execution from.
+ * @param[in] max_block Do not execute after this block.
+ * max_block may be executed, or the execution may stop earlier if the batch is full.
+ * @param[in] batch_size The size of DB changes to accumulate before returning from this method.
+ * Pass 0 if you want to execute just 1 block.
+ * @param[in] write_receipts Whether to write CBOR-encoded receipts into the DB.
+ *
+ * @param[out] last_executed_block The height of the last successfully executed block.
+ * Not written to if no blocks were executed, otherwise *last_executed_block ≤ max_block.
+ * @param[out] mdbx_error_code If an MDBX error occurs (this function returns kSilkwormMdbxError)
+ * and mdbx_error_code isn't NULL, it's populated with the relevant MDBX error code.
+ *
+ * @return A non-zero error value on failure and kSilkwormSuccess(=0) on success.
+ * kSilkwormBlockNotFound is probably OK: it simply means that the execution reached the end of the chain
+ * (blocks up to and incl. last_executed_block were still executed).
+ */
+SILKWORM_EXPORT
+enum SilkwormStatusCode silkworm_execute_blocks(
```

Comment:
It's probably good for start experimenting, but I suspect that the current Silkworm implementation is not exactly equivalent to the current Erigon implementation of the execution stage. Off the top of my head I can think of 2 discrepancies:
1.  Lack of support of Gnosis Chain & Polygon in Silkworm
2. No txpool notifications

Reply:
Yep, thanks for the heads up. This first iteration is a feasibility assessment, so I think we can postpone both topics for further development if the solution proves to be consistent and working at least for Ethereum w/o block production